Support parallelizing over collections that have non-Int lengths

sbromberger opened the pull request:
Changes `splitrange` to use `Integer` instead of `Int` so that mixed-width ints can be used. This fixes the following LightGraphs issue on 32-bit machines:
```
julia> g = Graph(10,20)
{10, 20} undirected simple Int32 graph

julia> Parallel.stress_centrality(g);

julia> g = Graph{Int64}(g)
{10, 20} undirected simple Int64 graph

julia> Parallel.stress_centrality(g);
ERROR: MethodError: no method matching splitrange(::Int64, ::Int32)
Closest candidates are:
  splitrange(::Int32, ::Int32) at /buildworker/worker/package_linuxarmv7l/build/usr/share/julia/stdlib/v1.0/Distributed/src/macros.jl:235
```

sbromberger committed:
Update macros.jl

Changes `splitrange` to use `Integer` instead of `Int` so that mixed-width ints can be used. This fixes the following LightGraphs issue on 32-bit machines:
```
julia> g = Graph(10,20)
{10, 20} undirected simple Int32 graph

julia> Parallel.stress_centrality(g);

julia> g = Graph{Int64}(g)
{10, 20} undirected simple Int64 graph

julia> Parallel.stress_centrality(g);
ERROR: MethodError: no method matching splitrange(::Int64, ::Int32)
Closest candidates are:
  splitrange(::Int32, ::Int32) at /buildworker/worker/package_linuxarmv7l/build/usr/share/julia/stdlib/v1.0/Distributed/src/macros.jl:235
```

sbromberger commented:
@pfitzseb on Discourse (I agree wholeheartedly):

> imho all functions there should use `Integer` or `T where T<:Integer`

sbromberger commented:
x-ref: https://github.com/JuliaGraphs/LightGraphs.jl/issues/985

JeffBezanson commented:
A slightly better fix is to change the calls to `splitrange` to `splitrange(Int(N), ...)`. Otherwise `splitrange` will be unnecessarily specialized. At least, the `np` argument is always an `Int` and doesn't need to be changed.

sbromberger commented:
@JeffBezanson - ref #28652. Thank you.

sbromberger commented:
I'll leave it to someone else to decide, then. My #26852 alternative was closed (I disagree with the closure, but whatever). In any rate, I've fixed it locally.

sbromberger commented:
Gah. Didn't mean to close this.

What path do we want to take?

sbromberger committed:
Update macros.jl

update per @jeffbezansons' suggestions

sbromberger commented:
Ping on this - any chance it could be merged soon? This fixes LightGraphs on 32-bit machines.

vchuravy reviewed (approved):
LGTM, would be great to have a test though.

sbromberger commented:
Could someone else please write the test? I edited this in GitHub's editor and can't figure out how to add a different file to the change set.

KristofferC commented:
Check the branch out locally or go to https://github.com/sbromberger/julia/tree/patch-1 and edit with GitHub.

sbromberger committed:
Update distributed_exec.jl

added test.

sbromberger committed:
Update distributed_exec.jl

Moved to before `addprocs` since we don't need remote workers for the `splitrange` test.

andreasnoack commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -15,6 +15,11 @@ include(joinpath(Sys.BINDIR, "..", "share", "julia", "test", "testenv.jl"))
         1
     end
 
+# PR #28651
```
Maybe
```julia
for T in (Int64, Int128)
  n = @distributed (+) for i in Base.OneTo(T(10))
    i
  end
  @test n == 55
end
```
This only happens whenever `length` returns something that is not an `Int` which is relatively rare.

sbromberger replied:
Yep. I had something written out last night but yours is simpler. Thanks.

sbromberger committed:
Update distributed_exec.jl

sbromberger committed:
Update distributed_exec.jl

sbromberger commented:
> This test doesn't test what was implemented (and thus fails).

Yup. The test was for the "old" fix. Sorry 'bout that. I've just pushed new tests (hat tip: @andreasnoack) that hopefully will work.

vchuravy commented:
Thanks Seth! Note to whoever merges this, please do a squash merge.

sbromberger commented:
> please do a squash merge.

It'd be better for my reputation if you did a fixup instead, but I'll take a squash :)

KristofferC commented:
The commit messages can be modified when squashing so all traces of Github editing can be removed ;)

sbromberger commented:
Travis failure is OSX-only and seems to be unrelated:
```
The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over).

The job has been terminated
```
